dockerTools: Refactor common writing code into writeImageStream

jhol opened the pull request:
Previously, the `buildLayeredImage` and `buildNixShellImage` functions
duplicated image compression functionality that was used to store the
output of `streamLayeredImage` and `streamNixShellImage`.
    
This patch factors out the common code into a combined function:
`writeImageStream`.
    
Doing this also makes it far more practical to customize and override
the behaviour of `streamLayeredImage` and `streamNixShellImage`, because
`buildLayeredImage` and `buildNixShellImage` formed an opaque wrapping
around the stream functions.
    
Therefore, this patch deprecates `buildLayeredImage` and
`buildNixShellImage`, and encourages users to write images by combining
one of the stream functions with `writeImageStream`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jhol commented on pkgs/build-support/docker/default.nix:
```diff
@@ -1412,21 +1406,38 @@ rec {
 
   # Wrapper around `streamNixShellImage` to build an image from the result.
   #
+  # Deprecated. Use `streamNixShellImage` + `writeImageStream` instead.
```
I'm on the fence about whether they should be deprecated at all.

While for my use-case, the `buildLayeredImage` and `buildNixShellImage` functions are problematic because they block out customization, is that a problem for the general audience?

Also, it would require the [demo on the nixos.org home page](https://nixos.org/#asciinema-demo-example_4) to be updated:.

If we go ahead with deprecation, we should certainly do it properly with the appropriate warnings.

roberth commented on doc/build-helpers/images/dockertools.section.md:
```diff
@@ -1508,6 +1516,81 @@ The environment in the image doesn't match `nix-shell` or `nix-build` exactly, a
+
+`writeImageStream` takes an executable that outputs an image stream (e.g. `streamLayeredImage`, `streamNixShellImage`) and writes a Docker-compatible repository tarball with optional compression.
+
+This is preferred over helpers that create an image stream internally and return its compressor derivation (e.g. `buildLayeredImage`, `buildNixShellImage`) because streams can be composed with overrides.
```
It's not entirely obvious what "compressor derivation" is.
Also let's pull those helpers to the front.

```suggestion
While helper functions are available that both configure the image and write it out, these do not support configuration overrides, so it may be preferable to use `writeImageStream` at the boundary of your expressions, while working with streaming derivations internally.
This way, you have more opportunities to override the image configurations.
```

roberth commented on doc/build-helpers/images/dockertools.section.md:
```diff
@@ -1508,6 +1516,81 @@ The environment in the image doesn't match `nix-shell` or `nix-build` exactly, a
+
+### Inputs {#ssec-pkgs-dockerTools-writeImageStream-inputs}
+
+`stream` (Path)
```
Not sure if it's confusing, but this is not the actual literal stream (although then it would have to be "captured" as a file anyway...).
Maybe `streamableImage` is an improvement?
This would then also apply to the passthru attr.

roberth reviewed (commented):
LGTM with some suggestions

jhol committed:
dockerTools: Refactor common writing code into writeImageStream

Previously, the buildLayeredImage and buildNixShellImage functions
duplicated image compression functionality that was used to store the
output of streamLayeredImage and streamNixShellImage.

This patch factors out the common code into a combined function:
writeImageStream.

Doing this also makes it far more practical to customize and override
the behaviour of streamLayeredImage and streamNixShellImage, because
buildLayeredImage and buildNixShellImage formed an opaque wrapping
around the stream functions.

Therefore, this patch adds a recommendation that users replace
buildLayeredImage and buildNixShellImage, and encourages users to write
images by combining one of the stream functions with writeImageStream.

Co-authored-by: commiterate <111539270+commiterate@users.noreply.github.com>
Co-authored-by: Robert Hensing <robert@roberthensing.nl>

jhol committed:
dockerTools: Refactor common writing code into writeImageStream

Previously, the buildLayeredImage and buildNixShellImage functions
duplicated image compression functionality that was used to store the
output of streamLayeredImage and streamNixShellImage.

This patch factors out the common code into a combined function:
writeImageStream.

Doing this also makes it far more practical to customize and override
the behaviour of streamLayeredImage and streamNixShellImage, because
buildLayeredImage and buildNixShellImage formed an opaque wrapping
around the stream functions.

Therefore, this patch adds a recommendation that users replace
buildLayeredImage and buildNixShellImage, and encourages users to write
images by combining one of the stream functions with writeImageStream.

Co-authored-by: commiterate <111539270+commiterate@users.noreply.github.com>
Co-authored-by: Robert Hensing <robert@roberthensing.nl>

jhol commented:
@roberth - This should be ready - notwithstanding the question of the naming of `stream`.

jhol commented:
@roberth Do you have time to take a final look at this? There is one outstanding discussion about the naming of the `stream` argument. I am happy to go with whatever you prefer on the naming of the argument.

jhol commented:
@roberth Just rebased. Any chance you could take a look? Thanks

jhol commented:
@roberth just rebased again. Please can you take another look? I think this would be a great improvement to the module. Or please can you let me know if there's any flaws you see in the conecpt?
